junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java

```diff
@@ -1479,12 +1560,102 @@ private static void failArraysNotEqual(Object expected, Object actual, Deque<Int
-		if (indexes == null || indexes.isEmpty()) {
-			return "";
+	// --- assertIterableEquals helpers ----------------------------------
+
+	private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+
+		if (expected == actual) {
+			return;
 		}
-		String indexesString = indexes.stream().map(Object::toString).collect(joining("][", "[", "]"));
-		return " at index " + indexesString;
+		assertIterablesNotNull(expected, actual, indexes, messageSupplier);
+
+		Iterator<?> expectedIterator = expected.iterator();
+		Iterator<?> actualIterator = actual.iterator();
+
+		int processed;
+		for (processed = 0; expectedIterator.hasNext() && actualIterator.hasNext(); processed++) {
+			Object expectedElement = expectedIterator.next();
+			Object actualElement = actualIterator.next();
+
+			if (expectedElement == actualElement) {
+				continue;
+			}
+
+			indexes.addLast(processed);
+			assertIterableElementsEqual(expectedElement, actualElement, indexes, messageSupplier);
+			indexes.removeLast();
+		}
+
+		assertIteratorsAreEmpty(expectedIterator, actualIterator, processed, indexes, messageSupplier);
+	}
+
+	private static void assertIterableElementsEqual(Object expected, Object actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+		if (expected instanceof Iterable && actual instanceof Iterable) {
+			assertIterableEquals((Iterable<?>) expected, (Iterable<?>) actual, indexes, messageSupplier);
+		}
+		else if (!Objects.equals(expected, actual)) {
+			if (expected == null && actual instanceof Iterable) {
+				failExpectedIterableIsNull(indexes, messageSupplier);
+			}
+			else if (expected instanceof Iterable && actual == null) {
+				failActualIterableIsNull(indexes, messageSupplier);
+			}
+			else {
+				failIterablesNotEqual(expected, actual, indexes, messageSupplier);
+			}
+		}
+	}
+
+	private static void assertIterablesNotNull(Object expected, Object actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+
+		if (expected == null) {
+			failExpectedIterableIsNull(indexes, messageSupplier);
+		}
+		if (actual == null) {
+			failActualIterableIsNull(indexes, messageSupplier);
+		}
+	}
+
+	private static void failExpectedIterableIsNull(Deque<Integer> indexes, Supplier<String> messageSupplier) {
+		fail(buildPrefix(nullSafeGet(messageSupplier)) + "expected iterable was <null>" + formatIndexes(indexes));
+	}
+
+	private static void failActualIterableIsNull(Deque<Integer> indexes, Supplier<String> messageSupplier) {
+		fail(buildPrefix(nullSafeGet(messageSupplier)) + "actual iterable was <null>" + formatIndexes(indexes));
+	}
+
+	private static void assertIteratorsAreEmpty(Iterator<?> expected, Iterator<?> actual, int processed,
+			Deque<Integer> indexes, Supplier<String> messageSupplier) {
+
+		if (expected.hasNext() || actual.hasNext()) {
+			int expectedCount = processed;
+			while (expected.hasNext()) {
```

Comment:
I thought about that, but it doesn't compile because it forces `expectedCount` to be final or effectively final, meaning I'd have to wrap expectedCount in an `int[]`, which in my opinion is less clean a solution than the current one.